### ISSUE TITLE
Flame createModule - Problems with unicode

### ISSUE THREAD

**Issue by alexsilva** [irmen/Pyro4#79]
Python 2.7 x64
Pyro4

Client Lua
https://github.com/alexsilva/luaPyro

Client code:

```
local python = plugin.Python:new()
local os = python.module({'os'})
local lua_mod = python.sendmodule{'lua.mod', [[
def openfile(filename, data):
    with open(filename, 'w') as f:
        f.write(data)

def getfile(filename):
    with open(filename, 'r') as f:
        return f.read()
]]}
```

You can not send a module whose name is unicode (who generated a unicode name was Python).

```
traceback: {
  1: Traceback (most recent call last):
  2: File "...\Pyro4\core.py", line 966, in handleRequest data = method(*vargs, **kwargs) # this is the   actual method call to the Pyro object
  3: File "...\Pyro4\utils\flame.py", line 226, in sendmodule createModule(modulename, modulesource)
  4: File "...\Pyro4\utils\flame.py", line 289, in createModule setattr(module, component, types.ModuleType(real_path))
  5: TypeError: module.__init__() argument 1 must be string, not unicode
}
args: {
  1: module.__init__() argument 1 must be string, not unicode
}
class: exceptions.TypeError
```

Memory dump (components)

```
components = {list} [u'lua', u'mod']
filename = {str} '<dynamic-module>'
module = {module} <module 'pyro-flame-module-context' (built-in)>
name = {unicode} u'lua.mod'
namespace = {NoneType} None
path = {str} ''
source = {unicode} u'\ndef openfile(filename, data):\n    with open(filename, \\'w\\') as f:\n        for line in data:\n            f.write(line)\n\ndef getfile(filename):\n    with open(filename, \\'w\\') as f:\n        return f.read()\n'
```

Error location
https://github.com/irmen/Pyro4/blob/master/src/Pyro4/utils/flame.py#L289

**Comment by alexsilva** [comment]
Fix to Python 2.7 x64

```
name = name.encode('utf-8')
```

**Comment by irmen** [comment]
Pyro (or rather, Flame) is expecting the client code to conform to the API that it exposes there. In this case, it is the module API of Python. If it doesn't accept unicode, *don't send unicode*.  It is not Pyro's task to fix up bad input before calling into python.

Another solution is to use Python 3.x instead of Python 2.

**Comment by alexsilva** [comment]
My client does not send unicode Lua, Lua 3.2 not know what unicode, this was a bug generated by Python to convert the JSON, do not fix this is to accept that your library has one that a bug and it will not be fixed!

**Comment by irmen** [comment]
What JSON?  Flame only works with the Pickle serializer.

Please show the code that you run on the server, the client code, and the full pyro logfile on the server (debug+wire log on).

**Comment by irmen** [comment]
Also, if you expect this to work with JSON, it won't. Pyro uses the standard json library from python itself and it always decodes all strings into unicode;

```
$ py -2.7
Python 2.7.9 (default, Dec 10 2014, 12:24:55) [MSC v.1500 32 bit (Intel)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import json
>>> json.loads('{"xxx": 555}')
{u'xxx': 555}
>>>
```

Use python 3.x if you want it to accept unicode names.

**Comment by irmen** [comment]
Closing this because it seems you're trying to work around the limitations of Flame as they are (and they won't be changed).

- Flame requires pickle serializer exclusively. This is not something that can/will be changed.
- The flameserver sets this correctly and won't accept other serilalizers, see https://github.com/irmen/Pyro4/blob/master/src/Pyro4/utils/flameserver.py#L40     I'm not sure how you are able to submit a json serialized message to the flame server without modifying this?
- Your particular problem disappears when using python 3.

**Comment by alexsilva** [comment]
Understand my case. I use Pyro4 to add functionality that has no Lua 3.2.
I can only speak to the Python using JSON, this is a limitation of the Lua.
I am creating a plugin to an implemented system in Lua 3.2.

It is a mixture of Python 2.7, Django 1.7, Publique! CMS (Lua 3.2).


I had to make adaptations for flame working with Json

```
In the flame module:

class Flame(Pyro4.utils.flame.Flame):
    def __init__(self):
        pass

    def module(self, name):
        packages = name.split('.')

        package = __import__(packages[0])
        modules = [package] + packages[1:]

        sys.modules[name] = reduce(lambda a, b: getattr(a, b, a), modules)

        return Pyro4.utils.flame.FlameModule(self, name)


In serializer (Django types):

class JsonSerializer(Pyro4.util.JsonSerializer):
    name = 'json'

    def default(self, obj):
        if type(obj) is set:
            obj = list(obj)
        elif isinstance(obj, datetime.datetime) or isinstance(obj, datetime.date):
            obj = obj.isoformat()
        elif isinstance(obj, Model):
            obj = serializers.serialize('python', [obj])[0]
        elif isinstance(obj, ValuesQuerySet):
            obj = list(obj)
        elif isinstance(obj, QuerySet):
            obj = serializers.serialize('python', obj)
        else:
            obj = super(JsonSerializer, self).default(obj)
        return obj
```

The big problem is you took that flame can only work with pickle, but with the right implementation would be possible to work with any serializer.

I can not change the Python version now, because it is a complex system (Python with Django, and Lua 3.2 with Publique! CMS).

But as you will not want to change anything, once again I will have to adapt the code as below:

```
class Flame(Pyro4.utils.flame.Flame):
    def __init__(self):
        pass

    def module(self, name):
        packages = name.split('.')

        package = __import__(packages[0])
        modules = [package] + packages[1:]

        sys.modules[name] = reduce(lambda a, b: getattr(a, b, a), modules)

        return Pyro4.utils.flame.FlameModule(self, name)

    def sendmodule(self, modulename, modulesource):
        if isinstance(modulename, unicode):
            modulename = modulename.encode('utf-8')
        super(Flame, self).sendmodule(modulename, modulesource)
        return FlameModule(self, modulename)
```

**Comment by irmen** [comment]
Thanks for the explanation. Your solution is fine btw -- override the parts of Pyro/Flame that you need to customize.

Your statement that flame would work with any serializer is not true in general. Flame supports various operations and *for the general case* it is required to be able to serialize *all* (or rather, most) of the Python types. Only pickle is able to do this.  For certain cases maybe json or serpent is sufficient, but not for the general case.

It is documented that Flame requires pickle exclusively. If you want (or need) to work around this, you take the responsibility to create your own customizations for your particular use case in the way that you're doing now.
